-1

**Failed Tests**:   Build 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5bb56a/20862/summary.html
**COMMIT**: 8c10a1f620021063ffad8de860ffc458903c74a2
**CMSSW**: CMSSW_12_2_X_2021-11-29-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/35627/20862/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @bsunanda cms-sw/cmssw#36264
 - @kakwok cms-sw/cmssw#35836
 - @tsusa cms-sw/cmssw#36255
 - @mariadalfonso cms-sw/cmssw#36277
 - @bsunanda cms-sw/cmssw#36192
 - @bsunanda cms-sw/cmssw#36183
 - @dildick cms-sw/cmssw#35779
 - @wpcarvalho cms-sw/cmssw#36291
 - @mmusich cms-sw/cmssw#36283

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5bb56a/20862/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5bb56a/20862/git-merge-result


## Build

I found compilation error when building: 

<pre>>> Building LCG reflex dict from header file src/DataFormats/L1TMuon/src/classes.h
>> Compiling LCG dictionary: tmp/slc7_amd64_gcc900/src/DataFormats/L1TMuon/src/DataFormatsL1TMuon/a/DataFormatsL1TMuon_xr.cc
>> Building  shared library tmp/slc7_amd64_gcc900/src/DataFormats/L1TMuon/src/DataFormatsL1TMuon/libDataFormatsL1TMuon.so
Copying tmp/slc7_amd64_gcc900/src/DataFormats/L1TMuon/src/DataFormatsL1TMuon/libDataFormatsL1TMuon.so to productstore area:
>> Checking EDM Class Version for src/DataFormats/L1TMuon/src/classes_def.xml in libDataFormatsL1TMuon.so
error: class 'l1t::RegionalMuonShower' has a different checksum for ClassVersion 12. Increment ClassVersion to 13 and assign it to checksum 1591048325
Suggestion: You can run 'scram build updateclassversion' to generate src/DataFormats/L1TMuon/src/classes_def.xml.generated with updated ClassVersion
gmake: *** [tmp/slc7_amd64_gcc900/src/DataFormats/L1TMuon/src/DataFormatsL1TMuon/libDataFormatsL1TMuon.so] Error 1
Leaving library rule at DataFormats/L1TMuon
>> Leaving Package DataFormats/L1TMuon
>> Package DataFormats/L1TMuon built
</pre>

--- a/DQM/L1TMonitor/src/L1TStage2Shower.cc
+++ b/DQM/L1TMonitor/src/L1TStage2Shower.cc
@@ -166,8 +166,8 @@ void L1TStage2Shower::analyze(const edm::Event& e, const edm::EventSetup& c) {
     if (not Shower.isValid())
       continue;
     if (Shower.isOneNominalInTime() or Shower.isTwoLooseInTime() or Shower.isOneTightInTime()) {
-      int endcap = Shower.endcap();
-      int sector = Shower.sector();
+      int endcap = Shower.trackFinderType() == l1t::tftype::emtf_pos ? 1 : -1;
+      int sector = Shower.processor() + 1;
       if (Shower.isOneTightInTime())
         emtfShowerTypeOccupancy->Fill(sector, (endcap == 1) ? 7.5 : 0.5);
       if (Shower.isTwoLooseInTime())

--- a/DataFormats/L1TMuon/interface/RegionalMuonShower.h
+++ b/DataFormats/L1TMuon/interface/RegionalMuonShower.h
@@ -5,6 +5,8 @@
 #include "DataFormats/L1Trigger/interface/BXVector.h"
 #include "DataFormats/L1Trigger/interface/L1TObjComparison.h"
 
+#include "RegionalMuonCandFwd.h"  // For tftype.
+
 namespace l1t {
 
   class RegionalMuonShower;
@@ -31,9 +33,8 @@ namespace l1t {
     void setTwoLooseOutOfTime(const bool bit) { isTwoLooseOutOfTime_ = bit; }
     void setTwoLooseInTime(const bool bit) { isTwoLooseInTime_ = bit; }
 
-    void setEndcap(const int endcap) { endcap_ = endcap; }
-    void setSector(const unsigned sector) { sector_ = sector; }
-    void setLink(const int link) { link_ = link; };
+    /// Set the processor ID, track-finder type. From these two, the link is set
+    void setTFIdentifiers(int processor, tftype trackFinder);
 
     bool isValid() const;
     bool isOneNominalInTime() const { return isOneNominalInTime_; }
@@ -43,10 +44,12 @@ namespace l1t {
     bool isTwoLooseInTime() const { return isTwoLooseInTime_; }
     bool isTwoLooseOutOfTime() const { return isTwoLooseOutOfTime_; }
 
-    int endcap() const { return endcap_; }
-    int sector() const { return sector_; }
     /// Get link on which the MicroGMT receives the candidate
-    int link() const { return link_; }
+    const int link() const { return link_; };
+    /// Get processor ID on which the candidate was found (0..5 for OMTF/EMTF; 0..11 for BMTF)
+    const int processor() const { return processor_; };
+    /// Get track-finder which found the muon (bmtf, emtf_pos/emtf_neg or omtf_pos/omtf_neg)
+    const tftype trackFinderType() const { return trackFinder_; };
 
     bool operator==(const l1t::RegionalMuonShower& rhs) const;
     inline bool operator!=(const l1t::RegionalMuonShower& rhs) const { return !(operator==(rhs)); };
@@ -60,9 +63,9 @@ namespace l1t {
     bool isOneTightOutOfTime_;
     bool isTwoLooseInTime_;
     bool isTwoLooseOutOfTime_;
-    int endcap_;       //    +/-1.  For ME+ and ME-.
-    unsigned sector_;  //  1 -  6.
     int link_;
+    int processor_;
+    tftype trackFinder_;
   };
 
 }  // namespace l1t

--- a/DataFormats/L1TMuon/src/RegionalMuonShower.cc
+++ b/DataFormats/L1TMuon/src/RegionalMuonShower.cc
@@ -12,11 +12,32 @@ l1t::RegionalMuonShower::RegionalMuonShower(bool oneNominalInTime,
       isOneTightOutOfTime_(oneTightOutOfTime),
       isTwoLooseInTime_(twoLooseInTime),
       isTwoLooseOutOfTime_(twoLooseOutOfTime),
-      endcap_(0),
-      sector_(0),
-      link_(0) {}
+      link_(0),
+      processor_(0) {}
 
 l1t::RegionalMuonShower::~RegionalMuonShower() {}
+
+void l1t::RegionalMuonShower::setTFIdentifiers(int processor, tftype trackFinder) {
+  trackFinder_ = trackFinder;
+  processor_ = processor;
+
+  switch (trackFinder_) {
+    case tftype::emtf_pos:
+      link_ = processor_ + 36;  // range 36...41
+      break;
+    case tftype::omtf_pos:
+      link_ = processor_ + 42;  // range 42...47
+      break;
+    case tftype::bmtf:
+      link_ = processor_ + 48;  // range 48...59
+      break;
+    case tftype::omtf_neg:
+      link_ = processor_ + 60;  // range 60...65
+      break;
+    case tftype::emtf_neg:
+      link_ = processor_ + 66;  // range 66...71
+  }
+}
 
 bool l1t::RegionalMuonShower::isValid() const {
   return (isOneNominalInTime_ or isTwoLooseInTime_ or isOneTightInTime_);

--- a/DataFormats/L1TMuon/src/classes_def.xml
+++ b/DataFormats/L1TMuon/src/classes_def.xml
@@ -13,7 +13,8 @@
   <class name="l1t::RegionalMuonCandBxCollection"/>
   <class name="edm::Wrapper<l1t::RegionalMuonCandBxCollection>"/>
 
-  <class name="l1t::RegionalMuonShower" ClassVersion="11">
+  <class name="l1t::RegionalMuonShower" ClassVersion="12">
+    <version ClassVersion="12" checksum="1591048325"/>
     <version ClassVersion="11" checksum="376206249"/>
     <version ClassVersion="10" checksum="3501434665"/>
   </class>

--- a/EventFilter/L1TRawToDigi/plugins/implementations_stage2/CommonTokens.h
+++ b/EventFilter/L1TRawToDigi/plugins/implementations_stage2/CommonTokens.h
@@ -6,6 +6,7 @@
 #include "DataFormats/L1Trigger/interface/Jet.h"
 #include "DataFormats/L1Trigger/interface/Tau.h"
 #include "DataFormats/L1Trigger/interface/Muon.h"
+#include "DataFormats/L1Trigger/interface/MuonShower.h"
 
 #include "FWCore/Utilities/interface/EDGetToken.h"
 #include "EventFilter/L1TRawToDigi/interface/PackerTokens.h"
@@ -19,6 +20,7 @@ namespace l1t {
       inline const edm::EDGetTokenT<JetBxCollection>& getJetToken() const { return jetToken_; };
       inline const edm::EDGetTokenT<TauBxCollection>& getTauToken() const { return tauToken_; };
       inline const edm::EDGetTokenT<MuonBxCollection>& getMuonToken() const { return muonToken_; };
+      inline const edm::EDGetTokenT<MuonShowerBxCollection>& getMuonShowerToken() const { return muonShowerToken_; };
 
     protected:
       edm::EDGetTokenT<EGammaBxCollection> egammaToken_;
@@ -26,6 +28,7 @@ namespace l1t {
       edm::EDGetTokenT<JetBxCollection> jetToken_;
       edm::EDGetTokenT<TauBxCollection> tauToken_;
       edm::EDGetTokenT<MuonBxCollection> muonToken_;
+      edm::EDGetTokenT<MuonShowerBxCollection> muonShowerToken_;
     };
   }  // namespace stage2
 }  // namespace l1t

--- a/EventFilter/L1TRawToDigi/plugins/implementations_stage2/GMTCollections.cc
+++ b/EventFilter/L1TRawToDigi/plugins/implementations_stage2/GMTCollections.cc
@@ -19,6 +19,12 @@ namespace l1t {
       event_.put(std::move(imdMuonsEMTFPos_), "imdMuonsEMTFPos");
       event_.put(std::move(imdMuonsOMTFNeg_), "imdMuonsOMTFNeg");
       event_.put(std::move(imdMuonsOMTFPos_), "imdMuonsOMTFPos");
+
+      event_.put(std::move(regionalMuonShowersEMTF_), "EMTF");
+      event_.put(std::move(muonShowers_[0]), "MuonShower");
+      for (int i = 1; i < 6; ++i) {
+        event_.put(std::move(muonShowers_[i]), "MuonShowerCopy" + std::to_string(i));
+      }
     }
   }  // namespace stage2
 }  // namespace l1t

--- a/EventFilter/L1TRawToDigi/plugins/implementations_stage2/GMTCollections.h
+++ b/EventFilter/L1TRawToDigi/plugins/implementations_stage2/GMTCollections.h
@@ -4,6 +4,9 @@
 #include "DataFormats/L1TMuon/interface/RegionalMuonCand.h"
 #include "DataFormats/L1Trigger/interface/Muon.h"
 
+#include "DataFormats/L1TMuon/interface/RegionalMuonShower.h"
+#include "DataFormats/L1Trigger/interface/MuonShower.h"
+
 #include "L1TObjectCollections.h"
 
 #include <array>
@@ -27,9 +30,15 @@ namespace l1t {
             imdMuonsEMTFNeg_(std::make_unique<MuonBxCollection>(0, oFirstBx, oLastBx)),
             imdMuonsEMTFPos_(std::make_unique<MuonBxCollection>(0, oFirstBx, oLastBx)),
             imdMuonsOMTFNeg_(std::make_unique<MuonBxCollection>(0, oFirstBx, oLastBx)),
-            imdMuonsOMTFPos_(std::make_unique<MuonBxCollection>(0, oFirstBx, oLastBx)) {
+            imdMuonsOMTFPos_(std::make_unique<MuonBxCollection>(0, oFirstBx, oLastBx)),
+
+            regionalMuonShowersEMTF_(std::make_unique<RegionalMuonShowerBxCollection>(0, iFirstBx, iLastBx)),
+            muonShowers_() {
         std::generate(muons_.begin(), muons_.end(), [&oFirstBx, &oLastBx] {
           return std::make_unique<MuonBxCollection>(0, oFirstBx, oLastBx);
+        });
+        std::generate(muonShowers_.begin(), muonShowers_.end(), [&oFirstBx, &oLastBx] {
+          return std::make_unique<MuonShowerBxCollection>(0, oFirstBx, oLastBx);
         });
       };
 
@@ -45,6 +54,11 @@ namespace l1t {
       inline MuonBxCollection* getImdMuonsOMTFNeg() { return imdMuonsOMTFNeg_.get(); };
       inline MuonBxCollection* getImdMuonsOMTFPos() { return imdMuonsOMTFPos_.get(); };
 
+      inline RegionalMuonShowerBxCollection* getRegionalMuonShowersEMTF() { return regionalMuonShowersEMTF_.get(); };
+      inline MuonShowerBxCollection* getMuonShowers(const unsigned int copy) override {
+        return muonShowers_[copy].get();
+      };
+
     private:
       std::unique_ptr<RegionalMuonCandBxCollection> regionalMuonCandsBMTF_;
       std::unique_ptr<RegionalMuonCandBxCollection> regionalMuonCandsOMTF_;
@@ -55,6 +69,9 @@ namespace l1t {
       std::unique_ptr<MuonBxCollection> imdMuonsEMTFPos_;
       std::unique_ptr<MuonBxCollection> imdMuonsOMTFNeg_;
       std::unique_ptr<MuonBxCollection> imdMuonsOMTFPos_;
+
+      std::unique_ptr<RegionalMuonShowerBxCollection> regionalMuonShowersEMTF_;
+      std::array<std::unique_ptr<MuonShowerBxCollection>, 6> muonShowers_;
     };
   }  // namespace stage2
 }  // namespace l1t

--- a/EventFilter/L1TRawToDigi/plugins/implementations_stage2/GMTTokens.cc
+++ b/EventFilter/L1TRawToDigi/plugins/implementations_stage2/GMTTokens.cc
@@ -26,6 +26,12 @@ namespace l1t {
       imdMuonTokenEMTFPos_ = cc.consumes<MuonBxCollection>(imdEmtfPosTag);
       imdMuonTokenOMTFNeg_ = cc.consumes<MuonBxCollection>(imdOmtfNegTag);
       imdMuonTokenOMTFPos_ = cc.consumes<MuonBxCollection>(imdOmtfPosTag);
+
+      auto emtfShowerTag = cfg.getParameter<edm::InputTag>("EMTFShowerInputLabel");
+      auto showerTag = cfg.getParameter<edm::InputTag>("ShowerInputLabel");
+
+      regionalMuonShowerTokenEMTF_ = cc.consumes<RegionalMuonShowerBxCollection>(emtfShowerTag);
+      muonShowerToken_ = cc.consumes<MuonShowerBxCollection>(showerTag);
     }
   }  // namespace stage2
 }  // namespace l1t

--- a/EventFilter/L1TRawToDigi/plugins/implementations_stage2/GMTTokens.h
+++ b/EventFilter/L1TRawToDigi/plugins/implementations_stage2/GMTTokens.h
@@ -2,6 +2,7 @@
 #define GMTTokens_h
 
 #include "DataFormats/L1TMuon/interface/RegionalMuonCand.h"
+#include "DataFormats/L1TMuon/interface/RegionalMuonShower.h"
 #include "DataFormats/L1Trigger/interface/Muon.h"
 
 #include "CommonTokens.h"
@@ -27,6 +28,10 @@ namespace l1t {
       inline const edm::EDGetTokenT<MuonBxCollection>& getImdMuonTokenOMTFNeg() const { return imdMuonTokenOMTFNeg_; };
       inline const edm::EDGetTokenT<MuonBxCollection>& getImdMuonTokenOMTFPos() const { return imdMuonTokenOMTFPos_; };
 
+      inline const edm::EDGetTokenT<RegionalMuonShowerBxCollection>& getRegionalMuonShowerTokenEMTF() const {
+        return regionalMuonShowerTokenEMTF_;
+      };
+
     private:
       edm::EDGetTokenT<RegionalMuonCandBxCollection> regionalMuonCandTokenBMTF_;
       edm::EDGetTokenT<RegionalMuonCandBxCollection> regionalMuonCandTokenOMTF_;
@@ -36,6 +41,8 @@ namespace l1t {
       edm::EDGetTokenT<MuonBxCollection> imdMuonTokenEMTFPos_;
       edm::EDGetTokenT<MuonBxCollection> imdMuonTokenOMTFNeg_;
       edm::EDGetTokenT<MuonBxCollection> imdMuonTokenOMTFPos_;
+
+      edm::EDGetTokenT<RegionalMuonShowerBxCollection> regionalMuonShowerTokenEMTF_;
     };
   }  // namespace stage2
 }  // namespace l1t

--- a/EventFilter/L1TRawToDigi/plugins/implementations_stage2/L1TObjectCollections.h
+++ b/EventFilter/L1TRawToDigi/plugins/implementations_stage2/L1TObjectCollections.h
@@ -6,6 +6,7 @@
 #include "DataFormats/L1Trigger/interface/Jet.h"
 #include "DataFormats/L1Trigger/interface/Tau.h"
 #include "DataFormats/L1Trigger/interface/Muon.h"
+#include "DataFormats/L1Trigger/interface/MuonShower.h"
 
 #include "DataFormats/EcalDigi/interface/EcalDigiCollections.h"
 
@@ -19,6 +20,7 @@ namespace l1t {
       ~L1TObjectCollections() override;
 
       virtual MuonBxCollection* getMuons(const unsigned int copy) { return nullptr; }
+      virtual MuonShowerBxCollection* getMuonShowers(const unsigned int copy) { return nullptr; }
       virtual EGammaBxCollection* getEGammas(const unsigned int copy) { return nullptr; }  //= 0;
       virtual EtSumBxCollection* getEtSums(const unsigned int copy) { return nullptr; }
       virtual JetBxCollection* getJets(const unsigned int copy) { return nullptr; }

--- a/L1Trigger/L1TMuon/interface/RegionalMuonRawDigiTranslator.h
+++ b/L1Trigger/L1TMuon/interface/RegionalMuonRawDigiTranslator.h
@@ -2,6 +2,7 @@
 #define RegionalMuonRawDigiTranslator_h
 
 #include "DataFormats/L1TMuon/interface/RegionalMuonCand.h"
+#include "DataFormats/L1TMuon/interface/RegionalMuonShower.h"
 
 namespace l1t {
   class RegionalMuonRawDigiTranslator {
@@ -15,11 +16,16 @@ namespace l1t {
                                      bool useEmtfDisplacementInfo);
     static void fillRegionalMuonCand(
         RegionalMuonCand& mu, uint64_t dataword, int proc, tftype tf, bool isKbmtf, bool useEmtfDisplacementInfo);
+    static bool fillRegionalMuonShower(
+        RegionalMuonShower& muShower, std::vector<uint32_t> bxPayload, int proc, tftype tf, bool useEmtfShowers);
     static void generatePackedDataWords(const RegionalMuonCand& mu,
                                         uint32_t& raw_data_00_31,
                                         uint32_t& raw_data_32_63,
                                         bool isKbmtf,
                                         bool useEmtfDisplacementInfo);
+    static void generatePackedShowerPayload(const RegionalMuonShower& shower,
+                                            std::array<uint32_t, 6>& payload,
+                                            bool useEmtfShowers);
     static uint64_t generate64bitDataWord(const RegionalMuonCand& mu, bool isKbmtf, bool useEmtfDisplacementInfo);
     static int generateRawTrkAddress(const RegionalMuonCand&, bool isKalman);
 
@@ -45,6 +51,15 @@ namespace l1t {
     static constexpr unsigned emtfPtUnconstrainedShift_ = 20;
     static constexpr unsigned trackAddressMask_ = 0x1FFFFFFF;
     static constexpr unsigned trackAddressShift_ = 2;
+    static constexpr unsigned emtfShowerOneNominalFrame_ = 0;
+    static constexpr unsigned emtfShowerOneNominalMask_ = 0x3;
+    static constexpr unsigned emtfShowerOneNominalInTimeShift_ = 18;
+    static constexpr unsigned emtfShowerOneNominalOutOfTimeShift_ = 19;
+    static constexpr unsigned emtfShowerTwoLooseFrame_ = 2;
+    static constexpr unsigned emtfShowerTwoLooseMask_ = 0x3;
+    static constexpr unsigned emtfShowerTwoLooseInTimeShift_ = 18;
+    static constexpr unsigned emtfShowerTwoLooseOutOfTimeShift_ = 19;
+
     // relative shifts within track address
     static constexpr unsigned bmtfTrAddrSegSelMask_ = 0xF;
     static constexpr unsigned bmtfTrAddrSegSelShift_ = 21;

--- a/L1Trigger/L1TMuon/src/RegionalMuonRawDigiTranslator.cc
+++ b/L1Trigger/L1TMuon/src/RegionalMuonRawDigiTranslator.cc
@@ -114,6 +114,44 @@ void l1t::RegionalMuonRawDigiTranslator::fillRegionalMuonCand(RegionalMuonCand& 
                        useEmtfDisplacementInfo);
 }
 
+bool l1t::RegionalMuonRawDigiTranslator::fillRegionalMuonShower(
+    RegionalMuonShower& muShower, std::vector<uint32_t> bxPayload, int proc, tftype tf, bool useEmtfShowers) {
+  if (useEmtfShowers && (tf == emtf_pos || tf == emtf_neg)) {
+    muShower.setTFIdentifiers(proc, tf);
+
+    muShower.setOneNominalInTime(((bxPayload[emtfShowerOneNominalFrame_] >> emtfShowerOneNominalInTimeShift_) & 1) ==
+                                 1);
+    muShower.setOneNominalOutOfTime(
+        ((bxPayload[emtfShowerOneNominalFrame_] >> emtfShowerOneNominalOutOfTimeShift_) & 1) == 1);
+    muShower.setTwoLooseInTime(((bxPayload[emtfShowerTwoLooseFrame_] >> emtfShowerTwoLooseInTimeShift_) & 1) == 1);
+    muShower.setTwoLooseOutOfTime(((bxPayload[emtfShowerTwoLooseFrame_] >> emtfShowerTwoLooseOutOfTimeShift_) & 1) ==
+                                  1);
+
+    return true;
+  } else {
+    return false;
+  }
+}
+
+void l1t::RegionalMuonRawDigiTranslator::generatePackedShowerPayload(const RegionalMuonShower& shower,
+                                                                     std::array<uint32_t, 6>& payload,
+                                                                     const bool useEmtfShowers) {
+  if (!useEmtfShowers) {
+    return;
+  }
+  // First we check whether we're going to overwrite something in the payload.
+  if (((payload.at(emtfShowerOneNominalFrame_) & emtfShowerOneNominalMask_) != 0) ||
+      ((payload.at(emtfShowerTwoLooseFrame_) & emtfShowerTwoLooseMask_) != 0)) {
+    edm::LogError("L1T") << "Check constants for RegionalMuonShower fields! It looks like we're in danger of "
+                            "overwriting muon data in the packer!";
+    return;
+  }
+  payload.at(emtfShowerOneNominalFrame_) |= (shower.isOneNominalInTime() & 1) << emtfShowerOneNominalInTimeShift_ |
+                                            (shower.isOneNominalOutOfTime() & 1) << emtfShowerOneNominalOutOfTimeShift_;
+  payload.at(emtfShowerTwoLooseFrame_) |= (shower.isTwoLooseInTime() & 1) << emtfShowerTwoLooseInTimeShift_ |
+                                          (shower.isTwoLooseOutOfTime() & 1) << emtfShowerTwoLooseOutOfTimeShift_;
+}
+
 void l1t::RegionalMuonRawDigiTranslator::generatePackedDataWords(const RegionalMuonCand& mu,
                                                                  uint32_t& raw_data_00_31,
                                                                  uint32_t& raw_data_32_63,

--- a/L1Trigger/L1TMuonEndCap/interface/SectorProcessorShower.h
+++ b/L1Trigger/L1TMuonEndCap/interface/SectorProcessorShower.h
@@ -15,6 +15,8 @@
 #include "DataFormats/CSCDigi/interface/CSCShowerDigiCollection.h"
 #include "L1Trigger/L1TMuonEndCap/interface/Common.h"
 
+#include "DataFormats/L1TMuon/interface/RegionalMuonCandFwd.h"
+
 #include <vector>
 
 class SectorProcessorShower {

--- a/L1Trigger/L1TMuonEndCap/src/SectorProcessorShower.cc
+++ b/L1Trigger/L1TMuonEndCap/src/SectorProcessorShower.cc
@@ -67,10 +67,8 @@ void SectorProcessorShower::process(const CSCShowerDigiCollection& in_showers,
   if (accept) {
     // shower output
     l1t::RegionalMuonShower out_shower(hasOneNominalInTime, false, hasTwoLooseInTime, false, hasOneTightInTime, false);
-    // convert [1,2] to [1, -1]
-    const int endcap(endcap_ == 1 ? 1 : -1);
-    out_shower.setEndcap(endcap);
-    out_shower.setSector(sector_);
+    l1t::tftype tftype = (endcap_ == 1) ? l1t::tftype::emtf_pos : l1t::tftype::emtf_neg;
+    out_shower.setTFIdentifiers(sector_, tftype);
     out_showers.push_back(0, out_shower);
   }
 }

